flink-core/src/main/java/org/apache/flink/util/LinkedOptionalMapSerializer.java

```diff
@@ -0,0 +1,127 @@
+import org.apache.flink.util.function.BiConsumerWithException;
+import org.apache.flink.util.function.BiFunctionWithException;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * LinkedOptionalMapSerializer - A serializer of {@link LinkedOptionalMap}.
+ */
+@Internal
+public final class LinkedOptionalMapSerializer {
+
+	private static final long HEADER = 0x4f2c69a3d70L;
+
+	private LinkedOptionalMapSerializer() {
+	}
+
+	public static <K, V> void writeOptionalMap(
+		DataOutputView out,
+		LinkedOptionalMap<K, V> map,
+		BiConsumerWithException<DataOutputView, K, IOException> keyWriter,
+		BiConsumerWithException<DataOutputView, V, IOException> valueWriter) throws IOException {
+
+		out.writeLong(HEADER);
+		out.writeInt(map.size());
+		map.forEach(((keyName, key, value) -> {
+			out.writeUTF(keyName);
+
+			if (key == null) {
+				out.writeBoolean(false);
+			}
+			else {
+				out.writeBoolean(true);
+				writeFramed(out, keyWriter, key);
+			}
+
+			if (value == null) {
+				out.writeBoolean(false);
+			}
+			else {
+				out.writeBoolean(true);
+				writeFramed(out, valueWriter, value);
+			}
+		}));
+	}
+
+	public static <K, V> LinkedOptionalMap<K, V> readOptionalMap(
+		DataInputView in,
+		BiFunctionWithException<DataInputView, String, K, IOException> keyReader,
+		BiFunctionWithException<DataInputView, String, V, IOException> valueReader) throws IOException {
+
+		final long header = in.readLong();
+		checkState(header == HEADER, "Corrupted stream received header %d", header);
+
+		long mapSize = in.readInt();
+		LinkedOptionalMap<K, V> map = new LinkedOptionalMap<>();
+		for (int i = 0; i < mapSize; i++) {
+			String keyName = in.readUTF();
+
+			final K key;
+			if (in.readBoolean()) {
+				key = tryReadFrame(in, keyName, keyReader);
+			}
+			else {
+				key = null;
+			}
+
+			final V value;
+			if (in.readBoolean()) {
+				value = tryReadFrame(in, keyName, valueReader);
+			}
+			else {
+				value = null;
+			}
+
+			map.put(keyName, key, value);
+		}
+		return map;
+	}
+
+	private static <T> void writeFramed(DataOutputView out, BiConsumerWithException<DataOutputView, T, IOException> writter, T item) throws IOException {
+		ByteArrayOutputStreamWithPos streamWithPos = new ByteArrayOutputStreamWithPos();
+		try (DataOutputViewStreamWrapper inMemoryStream = new DataOutputViewStreamWrapper(streamWithPos)) {
+			writter.accept(inMemoryStream, item);
+		}
+		out.writeInt(streamWithPos.getPosition());
+		out.write(streamWithPos.getBuf(), 0, streamWithPos.getPosition());
+	}
+
+	@Nullable
+	private static <T> T tryReadFrame(DataInputView in, String keyName, BiFunctionWithException<DataInputView, String, T, IOException> reader) throws IOException {
+		final int bufferSize = in.readInt();
+		byte[] buf = new byte[bufferSize];
+		in.read(buf, 0, bufferSize);
+		DataInputDeserializer frame = new DataInputDeserializer(buf, 0, bufferSize);
```

Comment:
I don't think we need to set `startPos` / `endPos` here.
Just `new DataInputDeserializer(buf)` will do.